1b sentence embeddings (#146)
huggingface/blog: 1b-sentence-embeddings.md

| Cosine-similarity   | Dot-product |
|---------------------|-------------|
| Vector has highest similarity to itself since $cos(a, a)=1$.  |  Other vectors can have higher dot-products $dot(a, a) < dot (a, b)$. |
| With normalised vectors it is equal to the dot product. The max vector length is equals 1.  | It might be slower with certain approximate nearest neighbour methods since the max vector not known. |
| With normalised vectors, it is proportional to euclidian distance. It works with k-means clustering.  | It does not work with k-means clustering.  |

In practice, we used a scaled similarity because score differences tends to be to small and apply a scaling factor $C$ such that $sim_scaled(a, b) = C * sim(a, b) $ with typically $C = 20$ ([Henderson and al., 2020]([https://doi.org/10.18653/v1/2020.findings-emnlp.196), [Radford and al., 2021](http://proceedings.mlr.press/v139/radford21a.html)).

### Improving Quality with Better Batches

> In our method, we build batches of sample pairs $(a_i , p_i)$. We consider all other samples from the batch, $(a_i , p_j), i \neq j$, as negatives sample pairs. The batch composition is therefore a key training aspect. Given the literature in the domain, we particularly focused on three main aspects of the batch.

Reviewer: ```suggestion
In our method, we build batches of sample pairs $(a_i , p_i)$. We consider all other samples from the batch, $(a_i , p_j), i \neq j$, as negatives sample pairs. The batch composition is therefore a key training aspect. Given the literature in the domain, we mainly focused on three main aspects of the batch.
```

#### Size matters

In contrastive learning Larger batch size is synonymous with better performance. As shown in the Figure extracted from Qu and al., ([2021](https://doi.org/10.18653/v1/2021.naacl-main.466)), a larger batch size increases the results.

![snippet](assets/25_1b_sentence_embeddings/batch-size.png)

#### Hard Negatives

In the same figure, we observe that including hard negatives also improves performance. Hard negatives are sample $p_j$ which are hard to distinguish from $p_i$. In our example, it could be the pairs « What is the capital of France? » and « What is the capital of the US? » which have a close semantic content and requires precisely understand the full sentence to be answered. On the contrary, the samples  « What is the capital of France? » and «How many Star Wars movies is there?» are less difficult to distinguish since they do not refer to the same topic.

#### Cross dataset batches

> We concatenated multiple datasets to train our models. We built a large batch and gather samples from the same dataset in the batch in order to limit the topic distribution and favor hard negatives. However, we also mix at least two datasets in the batch, to learn a global structure between topics and not only a local structure within a topic.

Reviewer 2: ```suggestion
We concatenated multiple datasets to train our models. We built a large batch and gather samples from the same dataset in the batch in order to limit the topic distribution and favor hard negatives. However, we also mix at least two datasets in the batch to learn a global structure between topics and not only a local structure within a topic.
```

> We concatenated multiple datasets to train our models. We built a large batch and gather samples from the same dataset in the batch in order to limit the topic distribution and favor hard negatives. However, we also mix at least two datasets in the batch, to learn a global structure between topics and not only a local structure within a topic.

Reviewer: ```suggestion
We concatenated multiple datasets to train our models. We built a large batch and gathered samples from the same batch dataset to limit the topic distribution and favor hard negatives. However, we also mix at least two datasets in the batch to learn a global structure between topics and not only a local structure within a topic.
```

## Training infrastructure and data

> As mention earlier, the quantity of data and the batch size directly impact the model performances. As part of the project, we benefited from efficient hardware infrastructure. We trained our models on [TPUs](https://cloud.google.com/tpu) which are compute units developed by Google and super efficient for matrix multiplications. TPUs have some [hardware specificities](https://huggingface.co/docs/accelerate/quicktour.html#training-on-tpu) which might require some specific code implementation.

Reviewer 2: ```suggestion
As mentioned earlier, the quantity of data and the batch size directly impact the model performances. As part of the project, we benefited from efficient hardware infrastructure. We trained our models on [TPUs](https://cloud.google.com/tpu) which are compute units developed by Google and super efficient for matrix multiplications. TPUs have some [hardware specificities](https://huggingface.co/docs/accelerate/quicktour.html#training-on-tpu) which might require some specific code implementation.
```

Additionally, we trained models on a large corpus as we concatenated multiple datasets up to 1 billion sentence pairs! All datasets used are detailed for each model in the [model card](https://huggingface.co/flax-sentence-embeddings/all_datasets_v3_MiniLM-L12).

## Conclusion

> You can find all models and datasets we created during the challenge in our [HuggingFace repository](https://huggingface.co/flax-sentence-embeddings). We trained 20 general-purpose Sentence Transformers models such as Mini-LM ([Wang and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/3f5ee243547dee91fbd053c1c4a845aa-Abstract.html)), Roberta ([liu and al., 2019](https://arxiv.org/abs/1907.11692 )), DistilBERT ([Sanh and al., 2020](http://arxiv.org/abs/1910.01108)) or MPNet ([Song and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/c3a690be93aa602ee2dc0ccab5b7b67e-Abstract.html)). Our models Achieve SOTA on multiple general-purpose Sentence Similarity evaluation tasks. We also uploaded  [8 datasets](https://huggingface.co/flax-sentence-embeddings)  specialized for Question Answering, Sentence-Similarity, and Gender Evaluation. 

Reviewer 2: ```suggestion
You can find all models and datasets we created during the challenge in our [HuggingFace repository](https://huggingface.co/flax-sentence-embeddings). We trained 20 general-purpose Sentence Transformers models such as Mini-LM ([Wang and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/3f5ee243547dee91fbd053c1c4a845aa-Abstract.html)), Roberta ([liu and al., 2019](https://arxiv.org/abs/1907.11692 )), DistilBERT ([Sanh and al., 2020](http://arxiv.org/abs/1910.01108)) and MPNet ([Song and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/c3a690be93aa602ee2dc0ccab5b7b67e-Abstract.html)). Our models achieve SOTA on multiple general-purpose Sentence Similarity evaluation tasks. We also shared [8 datasets](https://huggingface.co/flax-sentence-embeddings) specialized for Question Answering, Sentence-Similarity, and Gender Evaluation. 
```

> You can find all models and datasets we created during the challenge in our [HuggingFace repository](https://huggingface.co/flax-sentence-embeddings). We trained 20 general-purpose Sentence Transformers models such as Mini-LM ([Wang and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/3f5ee243547dee91fbd053c1c4a845aa-Abstract.html)), Roberta ([liu and al., 2019](https://arxiv.org/abs/1907.11692 )), DistilBERT ([Sanh and al., 2020](http://arxiv.org/abs/1910.01108)) or MPNet ([Song and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/c3a690be93aa602ee2dc0ccab5b7b67e-Abstract.html)). Our models Achieve SOTA on multiple general-purpose Sentence Similarity evaluation tasks. We also uploaded  [8 datasets](https://huggingface.co/flax-sentence-embeddings)  specialized for Question Answering, Sentence-Similarity, and Gender Evaluation. 

Reviewer: ```suggestion
You can find all models and datasets we created during the challenge in our [HuggingFace repository](https://huggingface.co/flax-sentence-embeddings). We trained 20 general-purpose Sentence Transformers models such as Mini-LM ([Wang and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/3f5ee243547dee91fbd053c1c4a845aa-Abstract.html)), RoBERTa ([liu and al., 2019](https://arxiv.org/abs/1907.11692 )), DistilBERT ([Sanh and al., 2020](http://arxiv.org/abs/1910.01108)) or MPNet ([Song and al., 2020](https://proceedings.neurips.cc/paper/2020/hash/c3a690be93aa602ee2dc0ccab5b7b67e-Abstract.html)). Our models Achieve SOTA on multiple general-purpose Sentence Similarity evaluation tasks. We also uploaded  [8 datasets](https://huggingface.co/flax-sentence-embeddings)  specialized for Question Answering, Sentence-Similarity, and Gender Evaluation. 
```

> General sentence embeddings might be used for many applications. We proposed some [demonstration](https://huggingface.co/spaces/flax-sentence-embeddings/sentence-embeddings) for several applications:

Reviewer 2: It would be great if you mention Spaces here

```suggestion
General sentence embeddings might be used for many applications. We built a [Spaces demo](https://huggingface.co/spaces/flax-sentence-embeddings/sentence-embeddings) for several applications:
```

> * The **sentence similarity** module compares the similarity of the main text with other texts of your choice. In the background, we’ll create an embedding for each text, and then we’ll use the cosine similarity function to calculate a similarity metric between our main sentence and the others.

Reviewer 2: ```suggestion
* The **sentence similarity** module compares the similarity of the main text with other texts of your choice. In the background, the demo extracts the embedding for each text and computes the similarity between the source sentence and the other using cosine similarity.
```

> * **Asymmetric QAv compare the Answer likeliness of a given Query with answer candidates of your choice.

Reviewer 2: ```suggestion
* **Asymmetric QA compares the answer likeliness of a given query with answer candidates of your choice.
```

> * **Search / Clusterv return nearby answers from a query. For example, if you input « python », it will retrieve anything related in the sense of the dot-product. distance.

Reviewer 2: ```suggestion
* **Search / Cluster returns nearby answers from a query. For example, if you input « python », it will retrieve closest sentences using dot-product distance.
```

* **Gender Bias Evaluation** report *inherent gender bias* in training set via random sampling of the sentences. Given an anchor text without any mention of gender for target occupation and 2 propositions with gendered pronouns, we compare if models assign a higher similarity to a given proposition and therefore evaluate their proportion to favor a specific gender.

> The [Community week using JAX/Flax for NLP & CV](https://discuss.huggingface.co/t/open-to-the-community-community-week-using-jax-flax-for-nlp-cv/7104) has been an intense and highly rewarding experience! The quality of the actors’ interventions and their presence helped us all learn a lot. We hope all projects had as much fun as we did in ours. Whenever you have questions or suggestions, don’t hesitate to contact us!

Reviewer: I would clarify who the **actors'** are here, and swap out **guidance** (or something similar) for **intervention**.
